[TypeChecker] Obsolete and remove old diagnostics (CSDiag)

xedin opened the pull request:
Port last edge-case diagnostics and completely obsolete CSDiag:

* Diagnose mismatches related to use of `===` and `!==` via a tailored fix
* Diagnose ambiguities related to ephemeral pointers
* Detect and diagnose mismatches in single parameter function conversions
* Detect and diagnose OoO arguments in call to binary functions/operators
* Ignore attempt to bind type var to dependent member with incorrect base
* Don't try to infer closure types with destructured parameters or `ErrorExpr`s in the body
* Allow `simplifyRestrictedConstraintImpl` to diagnose contextual failures with optionals
* Increase impact of a missing conformance related to stdlib type
* Fix `unwrap optional base` fix to preserve l-valueness of a member type
* Validate presence of `_MaxBuiltinFloatType` before generating constraints for float literals

Resolves: rdar://problem/59066040

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Allow `simplifyRestrictedConstraintImpl` to diagnose contextual failures with optionals

Since `simplifyRestrictedConstraintImpl` has both parent types and
does nested type matching it's a good place to diagnose top-level
contextual problems like mismatches in underlying types of optionals.

xedin committed:
[Diagnostics] Diagnose mismatches related to use of `===` and `!==` via a tailored fix

xedin committed:
[CSFix] Teach `getStructuralTypeContext` about tuple initialization calls

xedin committed:
[ConstraintSystem] Increase impact of a missing conformance related to stdlib type

Prioritize type mismatches over conformance failures when stdlib
types are involved because it wouldn't be appropriate to suggest
to add such a conformance, so the problem is most likely related
to something else e.g. other overload choice has a better fix.

Consider following example:

```swift
struct S {
  init(_: Double) {}
  init<T: BinaryInteger>(_: T) {}
}

_ = S(Double("0"))
```

In cases like that it's better to prefer failable initializer
which takes a `String` and returns `Double?` and diagnose a
problem related to missing optional unwrap instead of missing
conformances related to a `String` argument of other `Double`
initializer just because it returns a concrete type.

xedin committed:
[Diagnostics] Diagnose ambiguities related to ephemeral pointers

Diagnose ambiguity related to overloaded declarations where only
*some* of the overload choices have ephemeral pointer warnings/errors
associated with them. Such situations have be handled specifically
because ephemeral fixes do not affect the score.

If all of the overloads have ephemeral fixes associated with them
it's much easier to diagnose through notes associated with each fix.

xedin committed:
[ConstraintSystem] Detect and diagnose mismatches in single parameter function conversions

When there is a conversion from e.g. `(A) -> Void` to `(B) -> Void`
matching between `A` and `B` is going to have a special locator which
doesn't end in `TupleElement`, so `repairFailures` has to account
for that and fix it just like regular argument mismatch.

Resolves: rdar://problem/59066040

xedin committed:
[ConstraintSystem] Ignore attempt to bind type var to dependent member with incorrect base

Just like in cases where both sides are dependent member types
with resolved base that can't be simplified to a concrete type
let's ignore this mismatch and mark affected type variable as a hole
because something else has to be fixed already for this to happen.

xedin committed:
[CSGen] Don't try to infer closure types with destructured parameters

If one of the parameters represents a destructured tuple
e.g. `{ (x: Int, (y: Int, z: Int)) in ... }` let's fail
inference and not attempt to solve the constraint system because:

a. Destructuring has already been diagnosed by the parser;
b. Body of the closure would have error expressions for
   each incorrect parameter reference and solver wouldn't
   be able to produce any viable solutions.

xedin committed:
[ConstraintSystem] Detect and diagnose OoO arguments in call to binary functions/operators

Identify problems like:

```swift
func foo(_ x: Int, _ y: String) {}

func bar(a: Int, b: String) {
  foo(b, a) // Instead of `foo(a, b)`
}
```

Where arguments are out-of-order and repair it by using OoO fix on the
parent locator.

xedin committed:
[CSDiag] NFC: Remove `visitApplyExpr` and associated logic

xedin committed:
[ConstraintSystem] NFC: Remove obsolete `CalleeCandidateInfo`

xedin committed:
[CSDiag] NFC: Move `diagnoseDeprecatedConditionalConformanceOuterAccess` to CSApply

xedin committed:
[ConstraintSystem] Fix `unwrap optional base` fix to preserve l-valueness of a member type

xedin committed:
[ConstraintSystem] Don't attempt to solve closures with error expressions inside

Problem(s) which caused `ErrorExpr` to be added to AST should be
diagnosed already and solver wouldn't be able to produce a viable
solution in such case anyway.

xedin committed:
[CSGen] Validate presence of `_MaxBuiltinFloatType` before generating constraints for float literals

xedin committed:
[ConstraintSystem] Remove now completely obsolete CSDiag

xedin committed:
[ConstraintSystem] NFC: Remove workarounds related to (now deprecated) CSDiag

xedin commented:
@swift-ci please test

hborla reviewed (approved):
🎉

xedin committed:
[ConstraintSystem] Let `simplifyRestrictedConstraintImpl` fix any failures related to arguments with restrictions

xedin commented:
@swift-ci please test

hamishknight commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2880,6 +2880,58 @@ static bool diagnoseConflictingGenericArguments(ConstraintSystem &cs,
+        anchor = cast<InOutExpr>(anchor)->getSubExpr();
+
+        auto overload = solution.getOverloadChoiceIfAvailable(
+            cs.getConstraintLocator(anchor));
```
What if the ambiguity is for the function rather than the argument? For example, something like:

```swift
func foo(@_nonEphemeral _ ptr: UnsafePointer<Int>) {}
func foo(_ ptr: UnsafeRawPointer) {}

func bar(_ x: inout Int) {
  foo(&x)
}
```

I think with this change we'll complain about `x` being ambiguous instead of `foo` being ambiguous.

I wonder if we can defer to `diagnoseAmbiguity` for this? In fact I think we might be able to change the `hasFixes` predicate here:

https://github.com/apple/swift/blob/352653910992aca045782389025c97fb82295aaf/lib/Sema/ConstraintSystem.cpp#L2616

such that rather than filtering out solutions with fixes, it filters out solutions with a non-zero fix score. This should allow us to fall through to `diagnoseAmbiguity` for both this case and more generally for solutions with only warning fixes.

What do you think?

xedin replied:
I was thinking about that before but I just don't want to add ad hoc checks in `salvage` and `diagnoseAmbiguityWithFixes` to find whether fix is ` TreatEphemeralAsNonEphemeral` so I have decided to do it all in one place, these are fixes after all... For situations you have described I believe argument is going to be the same declaration in all solutions which is a good indication that something else is ambiguous, I'm going add such example into the suite.

hamishknight reviewed (commented):
Yay!

xedin commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2880,6 +2880,58 @@ static bool diagnoseConflictingGenericArguments(ConstraintSystem &cs,
+
+      // Ephemeral pointers are only possible in argument positions.
+      if (auto *anchor = simplifyLocatorToAnchor(fix->getLocator())) {
+        anchor = cast<InOutExpr>(anchor)->getSubExpr();
```
Good point, for some reason I though that it would only be possible to get an ephemeral pointer if the argument is explicitly `InOutExpr`, I'm going to fix that...

xedin replied:
I think only possible situation when implicit `inout-to-pointer` is allowed is operator invocations where it's allowed to elide `&` in argument.

xedin committed:
[ConstraintSystem] Simplify `diagnoseAmbiguityWithEphemeralPointers`

Don't attempt to figure out what exactly is ambiguous, let
`diagnoseAmbiguity` take care of that. Simplify make sure
that only some of the solutions have fixes and these fixes
are all related to use of ephemeral pointers.

xedin commented:
@swift-ci please test

xedin commented:
🤞

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
:shipit:
